### PR TITLE
treewide: Modular setup hooks

### LINKED ISSUE #40046: Setup hooks are polluting NIX_LDFLAGS
I just realized that we don't have a good way to stop setup hooks from being run multiple times. For instance, this is the output for gpgme on my computer:

```
-lintl -lintl -lintl -lintl -lintl -lintl -lintl -lintl -lintl -lintl -L/nix/store/zl3vsrqfwx4xvxagdrdqav73yl53cxps-libc++-5.0.1/lib -lintl -L/nix/store/15jap16aibnk81idwlf37g91mx8akaqv-gpgme-1.11.1-dev/lib -lintl -lintl -L/nix/store/8rcmp881wmbh18mzrkd9kmjxh2w8cn5z-libgpg-error-1.28/lib -lintl -L/nix/store/dkmxfrzsi0bzg4a4b976ww08243fslbs-glib-2.56.0-dev/lib -lintl -L/nix/store/byjz3zsj2397x4bllp4axqcnc517jdl7-zlib-1.2.11-dev/lib -lintl -L/nix/store/553hwxk54h8z87m4xgga4n8dznd6n8vp-zlib-1.2.11/lib -lintl -L/nix/store/si0g5b52xhrpdd8xjd67xdpskfgfscid-libffi-3.2.1-dev/lib -lintl -L/nix/store/qm51pj3fsvvb34kl7s9p4vvis92vq3ml-libffi-3.2.1/lib -lintl -L/nix/store/s8yvb556vw7l73qzx2wmhc2qmdw972dy-gettext-0.19.8/lib -lintl -L/nix/store/6xismd1mcjqnsj873wmx1fr0p62v13f2-libiconv-osx-10.11.6/lib -lintl -L/nix/store/7cb7305cliycr62pkcc8yzpjl50jncad-glib-2.56.0/lib -lintl -lintl -L/nix/store/yr6dzd26vd5crjfczv5w5kpw0g3la7qg-libassuan-2.5.1/lib -lintl -L/nix/store/8ipd3lbwabrxwbm1f7sxi8xj54sn0a44-pth-2.0.7/lib -lintl -L/nix/store/6ilwnfqar3k2vx9baazzagbfh25xgcfg-gpgme-1.11.1/lib -lintl -lintl -lintl -lintl -lintl -lintl -lintl -lintl -lintl -lintl -lintl -lintl -L/nix/store/zl3vsrqfwx4xvxagdrdqav73yl53cxps-libc++-5.0.1/lib -lintl -L/nix/store/15jap16aibnk81idwlf37g91mx8akaqv-gpgme-1.11.1-dev/lib -lintl -lintl -L/nix/store/8rcmp881wmbh18mzrkd9kmjxh2w8cn5z-libgpg-error-1.28/lib -lintl -L/nix/store/dkmxfrzsi0bzg4a4b976ww08243fslbs-glib-2.56.0-dev/lib -lintl -L/nix/store/byjz3zsj2397x4bllp4axqcnc517jdl7-zlib-1.2.11-dev/lib -lintl -L/nix/store/553hwxk54h8z87m4xgga4n8dznd6n8vp-zlib-1.2.11/lib -lintl -L/nix/store/si0g5b52xhrpdd8xjd67xdpskfgfscid-libffi-3.2.1-dev/lib -lintl -L/nix/store/qm51pj3fsvvb34kl7s9p4vvis92vq3ml-libffi-3.2.1/lib -lintl -L/nix/store/s8yvb556vw7l73qzx2wmhc2qmdw972dy-gettext-0.19.8/lib -lintl -L/nix/store/6xismd1mcjqnsj873wmx1fr0p62v13f2-libiconv-osx-10.11.6/lib -lintl -L/nix/store/7cb7305cliycr62pkcc8yzpjl50jncad-glib-2.56.0/lib -lintl -lintl -L/nix/store/yr6dzd26vd5crjfczv5w5kpw0g3la7qg-libassuan-2.5.1/lib -lintl -L/nix/store/8ipd3lbwabrxwbm1f7sxi8xj54sn0a44-pth-2.0.7/lib -lintl -L/nix/store/6ilwnfqar3k2vx9baazzagbfh25xgcfg-gpgme-1.11.1/lib -lintl -lintl -lintl -lintl -lintl -lintl -lintl -lintl -lintl -lintl -lintl -lintl -L/nix/store/zl3vsrqfwx4xvxagdrdqav73yl53cxps-libc++-5.0.1/lib -lintl -L/nix/store/15jap16aibnk81idwlf37g91mx8akaqv-gpgme-1.11.1-dev/lib -lintl -lintl -L/nix/store/8rcmp881wmbh18mzrkd9kmjxh2w8cn5z-libgpg-error-1.28/lib -lintl -L/nix/store/dkmxfrzsi0bzg4a4b976ww08243fslbs-glib-2.56.0-dev/lib -lintl -L/nix/store/byjz3zsj2397x4bllp4axqcnc517jdl7-zlib-1.2.11-dev/lib -lintl -L/nix/store/553hwxk54h8z87m4xgga4n8dznd6n8vp-zlib-1.2.11/lib -lintl -L/nix/store/si0g5b52xhrpdd8xjd67xdpskfgfscid-libffi-3.2.1-dev/lib -lintl -L/nix/store/qm51pj3fsvvb34kl7s9p4vvis92vq3ml-libffi-3.2.1/lib -lintl -L/nix/store/s8yvb556vw7l73qzx2wmhc2qmdw972dy-gettext-0.19.8/lib -lintl -L/nix/store/6xismd1mcjqnsj873wmx1fr0p62v13f2-libiconv-osx-10.11.6/lib -lintl -L/nix/store/7cb7305cliycr62pkcc8yzpjl50jncad-glib-2.56.0/lib -lintl -lintl -L/nix/store/yr6dzd26vd5crjfczv5w5kpw0g3la7qg-libassuan-2.5.1/lib -lintl -L/nix/store/8ipd3lbwabrxwbm1f7sxi8xj54sn0a44-pth-2.0.7/lib -lintl -L/nix/store/6ilwnfqar3k2vx9baazzagbfh25xgcfg-gpgme-1.11.1/lib -lintl -lintl -lintl -lintl -lintl -lintl -lintl -lintl -lintl -lintl -lintl -lintl -L/nix/store/zl3vsrqfwx4xvxagdrdqav73yl53cxps-libc++-5.0.1/lib -lintl -L/nix/store/15jap16aibnk81idwlf37g91mx8akaqv-gpgme-1.11.1-dev/lib -lintl -lintl -L/nix/store/8rcmp881wmbh18mzrkd9kmjxh2w8cn5z-libgpg-error-1.28/lib -lintl -L/nix/store/dkmxfrzsi0bzg4a4b976ww08243fslbs-glib-2.56.0-dev/lib -lintl -L/nix/store/byjz3zsj2397x4bllp4axqcnc517jdl7-zlib-1.2.11-dev/lib -lintl -L/nix/store/553hwxk54h8z87m4xgga4n8dznd6n8vp-zlib-1.2.11/lib -lintl -L/nix/store/si0g5b52xhrpdd8xjd67xdpskfgfscid-libffi-3.2.1-dev/lib -lintl -L/nix/store/qm51pj3fsvvb34kl7s9p4vvis92vq3ml-libffi-3.2.1/lib -lintl -L/nix/store/s8yvb556vw7l73qzx2wmhc2qmdw972dy-gettext-0.19.8/lib -lintl -L/nix/store/6xismd1mcjqnsj873wmx1fr0p62v13f2-libiconv-osx-10.11.6/lib -lintl -L/nix/store/7cb7305cliycr62pkcc8yzpjl50jncad-glib-2.56.0/lib -lintl -lintl -L/nix/store/yr6dzd26vd5crjfczv5w5kpw0g3la7qg-libassuan-2.5.1/lib -lintl -L/nix/store/8ipd3lbwabrxwbm1f7sxi8xj54sn0a44-pth-2.0.7/lib -lintl -L/nix/store/6ilwnfqar3k2vx9baazzagbfh25xgcfg-gpgme-1.11.1/lib -lintl -lintl
```

I don't understand why the setup hook is being run so many times, but it is definitely not good. Is there a way to ensure ```addEnvHooks``` is only run once? Otherwise we may need to add a flag like ```HAS_ADDED_LINTL```.

Current packages adding ldflags:

- libiconv
- gettext (libintl)
- musl-fts
- netbsd.compat

/cc @ericson2314 @dtzWill

### PR DESCRIPTION
**Pull request by Ericson2314**
###### Motivation for this change

It's error prone to copy paste this "role" stuff everywhere. Also, the more similar the `*-wrapper`s are, the easier it will be add a `pkgconfig-wrapper` as @bgamari once mentioned needing. Also some of the env hooks should just be setup hooks (setup hooks directly rather than setup hooks that define env hooks).

I don't really like this `setupHooks` bloat, but after https://github.com/NixOS/nixpkgs/pull/31414 we can just make these helper setup hooks dependencies, which will work fine. I'd have loved to land that PR first, but fixing it requires https://github.com/NixOS/nixpkgs/issues/26004 which would be easier (but still hard) with this, so I do this stop-gap first.

Fixes #40046

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/development/libraries/libiconv/setup-hook.sh [hunk 1]
```diff
@@ -1,18 +1,6 @@
...
-        0)  local role='' ;;
-        1)  local role='TARGET_' ;;
-        *)  echo "cc-wrapper: Error: Cannot be used with $depHostOffset-offset deps" >2;
-            return 1 ;;
-    esac
 
-    export NIX_${role}LDFLAGS+=" -liconv"
-}
-
-addEnvHooks "$hostOffset" iconvLdflags
```
Comment (matthewbauer) [inline pkgs/development/libraries/libiconv/setup-hook.sh]: So were addEnvHooks ever needed here? I was not really sure what they were doing but it was used in many places.

File: pkgs/build-support/bintools-wrapper/default.nix [hunk 2]
```diff
@@ -300,7 +305,7 @@ stdenv.mkDerivation {
 
   inherit dynamicLinker expand-response-params;
 
-  # for substitution in utils.sh
+  # for substitution in utils.bash
```
Comment (matthewbauer) [inline pkgs/build-support/bintools-wrapper/default.nix]: What is making these bash scripts? I know we have in the past assumed bash - but it seemed nice to make these in theory work on non-Bash shells.

File: pkgs/development/libraries/libiconv/setup-hook.sh (hunk 1 above)
Comment (Ericson2314) [reply pkgs/development/libraries/libiconv/setup-hook.sh]: Nope! and env hook does something *per dependency*. It's a way to accumulate a bunch of things which are for-eached at the end. Basically, If your hook doesn't use `$1`, it shouldn't be an env hook.

File: pkgs/build-support/bintools-wrapper/default.nix (hunk 2 above)
Comment (Ericson2314) [reply pkgs/build-support/bintools-wrapper/default.nix]: Virtually all our stuff is bash-only. Given it's not MS-DOS with max-3 file extensions, I figure it's best to be honest on use `.bash` not `.sh` :D. More seriously, editors highlight better with `.bash`: emacs will at least check the shebang, but many of these hook files don't have one.

**Comment by matthewbauer** [comment]
> Fixes #40134

I think you have the wrong number there.

**Comment by Ericson2314** [comment]
fixed: #40046

File: pkgs/build-support/bintools-wrapper/default.nix [hunk 3]
```diff
@@ -190,7 +190,12 @@ stdenv.mkDerivation {
 
   depsTargetTargetPropagated = extraPackages;
 
-  setupHook = ./setup-hook.sh;
+  wrapperName = "BINTOOLS_WRAPPER";
+
+  setupHooks = [
+    ../setup-hooks/role.bash
+    ./setup-hook.sh
```
[the remark this answers is not in the document]
Comment (Ericson2314) [reply pkgs/build-support/bintools-wrapper/default.nix]: @matthewbauer well it still needs to be part of the derivation which means extra hoops anyways.

**Comment by matthewbauer** [comment]
Awesome work!

**Comment by matthewbauer** [comment]
Also might want to look to see if this fixes #40013. We also have a hack for inkscape having long args in a276d5160cc4b307d97261b419e183fb3a2214aa...047c9372ab1314d8e3fab13b9f21359fdf3a643c that may no longer be needed.

**Comment by matthewbauer** [comment]
@Ericson2314 
Just a heads up but this is giving an issue in staging:

https://hydra.nixos.org/build/74502711/nixlog/1

```
/nix/store/9i0zrfqzfcx639s20x95r10fv9j7v6id-libiconv-osx-10.11.6/nix-support/setup-hook: line 5: getHostRole: command not found
builder for '/nix/store/rcfli01dwjbzvbwhw5k1lpkwzh57rqky-cairo-1.15.10.drv' failed with exit code 127
```

Not sure if it's since been fixed or not.

**Comment by matthewbauer** [comment]
Hopefully fixed in 89e196d971e167867318c8fc2e08cb0782c75bec.

File: pkgs/stdenv/generic/setup.sh [hunk 4]
```diff
@@ -685,39 +672,72 @@ substitute() {
...
 
             *)
-                echo "substitute(): ERROR: Invalid command line argument: $1" >&2
+                echo "substituteStream(): ERROR: Invalid command line argument: $1" >&2
                 return 1
                 ;;
         esac
 
-        content="${content//"$pattern"/$replacement}"
+        eval "$var"'=${'"$var"'//"$pattern"/"$replacement"}'
```
Comment (Profpatsch) [inline pkgs/stdenv/generic/setup.sh]: What’s this commit do? Such major changes to the internals of `stdenv` should *at least* have an extensive commit message of all the implications. I’m told `substituteInPlace` is very slow (in the order of seconds), so on inquiring this is the last commit that popped up, with no commit description whatsoever. :(

Comment (dtzWill) [reply pkgs/stdenv/generic/setup.sh]: Was this resolved? Does this still have a performance problem? (If so, is there a known way to reproduce it?)

Comment (Profpatsch) [reply pkgs/stdenv/generic/setup.sh]: To reproduce it, try changing something in `pkgs/development/tools/build-managers/bazel/default.nix` and then run `nix-build -A bazel /path/to/nixpkgs`.

You should see that the substitution of every file takes like half a second, which is a lot for a simple text substitution.

See also https://github.com/NixOS/nixpkgs/pull/35304#issuecomment-445550010, cc @mboes

Comment (Ericson2314) [reply pkgs/stdenv/generic/setup.sh]: It's always been O(size of file * number of @var@ definitions), which is quiet shitty.

Just as we did for "expand response params", we should probably rewrite this in C or something.

**Comment by dtzWill** [comment]
I'm not seeing 0.5s locally, but it sure does add up!

As a workaround for this particular case,
I looked into it a bit and put together this workaround:

https://github.com/dtzWill/nixpkgs/commit/9ed2c64653d4ac45d5978f1926b32c1ebde967c3

LMK if this looks good (dunno if it's any trouble to try it),
if so we can submit a PR with it.

**Pull request merged**